lib/crypto/client.cpp

```diff
@@ -378,6 +378,94 @@ OlmClient::create_outbound_session(const std::string &identity_key, const std::s
         return session;
 }
 
+SASPtr
+OlmClient::sas_init()
+{
+        auto sas        = create_olm_object<SASObject>();
+        auto random_buf = create_buffer(olm_create_sas_random_length(sas.get()));
+
+        const int ret = olm_create_sas(sas.get(), random_buf.data(), random_buf.size());
+
+        if (ret == -1)
+                throw olm_exception("create_sas_instance", sas.get());
+
+        return sas;
+}
+
+std::string
+OlmClient::sas_get_pub_key(OlmSAS *sas)
+{
+        auto pub_key_buffer = create_buffer(olm_sas_pubkey_length(sas));
+
+        const int ret = olm_sas_get_pubkey(sas, pub_key_buffer.data(), pub_key_buffer.size());
+
+        if (ret == -1)
+                throw olm_exception("get_public_key", sas);
+
+        return to_string(pub_key_buffer);
+}
+
+void
+OlmClient::set_their_key(OlmSAS *sas, std::string their_public_key)
+{
+        auto pub_key_buffer = to_binary_buf(their_public_key);
+
+        const int ret = olm_sas_set_their_key(sas, pub_key_buffer.data(), pub_key_buffer.size());
+
+        if (ret == -1)
+                throw olm_exception("get_public_key", sas);
+}
+
+std::vector<int>
+OlmClient::generate_bytes(OlmSAS *sas, std::string info, mtx::events::msg::SASMethods method)
```

Comment:
Not so sure about this. I think, if you are already returning different things depending on the method, you should probably return the emojis or the number depending on the method. At which point it would already be easier to have 2 separate methods or just letting nheko do it. I think it actually may already do that partially.